src/Microsoft.AspNet.Http.Abstractions/HttpResponse.cs

```diff
@@ -24,14 +25,33 @@ public abstract class HttpResponse
 
         public abstract bool HasStarted { get; }
 
-        public abstract void OnResponseStarting(Action<object> callback, object state);
+        public abstract void OnResponseStarting(Func<object, Task> callback, object state);
 
-        public abstract void OnResponseCompleted(Action<object> callback, object state);
+        public virtual void OnResponseStarting(Func<Task> callback) => OnResponseStarting(s => callback(), state: null);
 
-        public virtual void Redirect(string location)
-        {
-            Redirect(location, permanent: false);
-        }
+        public virtual void OnResponseStarting(Action callback) => OnResponseStarting(s => callback(), state: null);
+
+        public virtual void OnResponseStarting(Action<object> callback, object state) =>
```

Comment:
Does the intellisense get confused with anonymous delegates between Action and Func?